elasticapm/utils/cloud.py

```diff
@@ -0,0 +1,161 @@
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import os
+import socket
+
+import urllib3
+
+
+def aws_metadata():
+    """
+    Fetch AWS metadata from the local metadata server. If metadata server is
+    not found, return an empty dictionary
+    """
+    ret = {}
+    http = urllib3.PoolManager()
+
+    try:
+        # This will throw an error if the metadata server isn't available,
+        # and will be quiet in the logs, unlike urllib3
+        socket.create_connection(("169.254.169.254", 80), 0.1)
+
+        ttl_header = {"X-aws-ec2-metadata-token-ttl-seconds": "300"}
+        token_url = "http://169.254.169.254/latest/api/token"
+        token_request = http.request("PUT", token_url, headers=ttl_header, timeout=3.0)
+        token = token_request.data.decode("utf-8")
+        aws_token_header = {"X-aws-ec2-metadata-token": token} if token else {}
+        metadata = json.loads(
+            http.request(
+                "GET",
+                "http://169.254.169.254/latest/dynamic/instance-identity/document",
+                headers=aws_token_header,
+                timeout=3.0,
+                retries=False,
+            ).data.decode("utf-8")
+        )
+
+        ret = {
+            "account": {"id": metadata["accountId"]},
+            "instance": {"id": metadata["instanceId"]},
+            "availability_zone": metadata["availabilityZone"],
+            "machine": {"type": metadata["instanceType"]},
+            "provider": "aws",
+            "region": metadata["region"],
+        }
+    except Exception:
+        # Not on an AWS box
+        return {}
+    return ret
+
+
+def gcp_metadata():
+    """
+    Fetch GCP metadata from the local metadata server. If metadata server is
+    not found, return an empty dictionary
+    """
+    ret = {}
+    headers = {"Metadata-Flavor": "Google"}
+    http = urllib3.PoolManager()
+
+    try:
+        # This will throw an error if the metadata server isn't available,
+        # and will be quiet in the logs, unlike urllib3
+        socket.create_connection(("metadata.google.internal", 80), 0.1)
```

Comment:
Would switch to `socket.getaddrinfo("metadata.google.internal", 80, 0, SOCK_STREAM)` so the TCP handshake can be skipped. `create_connection()` uses the same method internally.